pkg/controller/node/cidr_allocator.go

```diff
@@ -0,0 +1,182 @@
+}
+
+// NewCIDRRangeAllocator return a CIDRAllocator to allocate CIDR for node
+// If subNetMaskSize is less than cluster CIDR mask size, an emptyAllocator is returned.
+// Otherwise an rangeAllocator is returned.
+func NewCIDRRangeAllocator(clusterCIDR *net.IPNet, subNetMaskSize int) CIDRAllocator {
+	cidrMask := clusterCIDR.Mask
+	cidrMaskSize, _ := cidrMask.Size()
+
+	if subNetMaskSize < cidrMaskSize {
+		return &emptyAllocator{}
+	}
+
+	ra := &rangeAllocator{
+		clusterCIDR:    clusterCIDR,
+		subNetMaskSize: subNetMaskSize,
+		cidrs:          make(chan *net.IPNet, cidrsBufSize),
+		rand:           rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+	go util.Until(func() { ra.fillCIDRs(util.NeverStop) }, nextFreeCIDRCooldown, util.NeverStop)
+	return ra
+}
+
+func (r *rangeAllocator) AllocateNext() (cidr *net.IPNet, err error) {
+	select {
+	case cidr = <-r.cidrs:
+	case <-time.After(allocateNextTimeout):
+		err = errCIDRRangeNoCIDRsRemaining
+	}
+	return
+}
+
+func (r *rangeAllocator) Release(cidr *net.IPNet) {
+	clusterIP := r.clusterCIDR.IP.To4()
+	clusterMask := r.clusterCIDR.Mask
+	clusterMaskSize, _ := clusterMask.Size()
+
+	nodeNum := uint32(1) << uint32(r.subNetMaskSize-clusterMaskSize)
+
+	used := (binary.BigEndian.Uint32(clusterIP) ^ binary.BigEndian.Uint32(cidr.IP.To4())) >> uint32(32-r.subNetMaskSize)
+
+	if used > nodeNum {
+		return
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.used.SetBit(&r.used, int(used), 0)
+}
+
+// fillCIDRs loops, always searching for the next free CIDR and, if found, fills the CIDRs buffer with it.
+// this func blocks until either there are no remaining free CIDRs, or else the stopCh chan is closed.
+func (r *rangeAllocator) fillCIDRs(stopCh <-chan struct{}) {
+	for {
+		cidr := r.nextFreeCIDR()
+		if cidr == nil {
+			return
+		}
+		select {
+		case <-stopCh:
+			return
+		case r.cidrs <- cidr:
+		}
+	}
+}
+
+// nextFreeCIDR finds a free CIDR, first picking a random cidr, if it is already in use
+// then the CIDR range is scanned sequentially until either a CIDR is found or the scan completes
+// unsuccessfully. an unsuccessful scan returns a nil.
+func (r *rangeAllocator) nextFreeCIDR() *net.IPNet {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	clusterIP := r.clusterCIDR.IP.To4()
+	clusterMask := r.clusterCIDR.Mask
+	clusterMaskSize, _ := clusterMask.Size()
+
+	nodeNum := 1 << uint32(r.subNetMaskSize-clusterMaskSize)
+
+	nextUnused, err := func() (int, error) {
+		j := r.rand.Intn(nodeNum)
+		if r.used.Bit(j) == 0 {
+			return j, nil
+		}
+		// search sequentially
+		for i := j + 1; i < nodeNum; i++ {
+			if r.used.Bit(i) == 0 {
+				return i, nil
+			}
+		}
+		for i := 0; i < j; i++ {
+			if r.used.Bit(i) == 0 {
+				return i, nil
+			}
+		}
+		return -1, fmt.Errorf("not found")
+	}()
+
+	if err != nil {
+		return nil
+	}
+	r.used.SetBit(&r.used, nextUnused, 1)
+
+	ip := make([]byte, 4)
+	j := uint32(nextUnused) << uint32(32-r.subNetMaskSize)
+	ipInt := (binary.BigEndian.Uint32(clusterIP)) | j
```

Comment:
This would give incorrect (duplicate) results if clusterCIDR is not actually the first address in the range. That kind of mistake shouldn't happen, but someone, somewhere will make a typo. I suppose this should be validated upstream in controllermanager.go and log a warning (while ANDing the IP with the mask) or terminate execution.